Option to split during conversion

christianazinn opened the pull request:
This PR introduces additional options to `convert.py` that allow users to split a model into shards while converting rather than having to do it after conversion, including a default small first shard as outlined in #6463.

Other functionality we ought to have includes `--split-max-size` (so far it's just `--split-max-tensors`), displaying estimated shard sizes, dry running, and adding sharding for the other `convert-*-to-*.py` scripts. This will be considered a draft until those are worked out. Also needs considerable testing, but luckily as this deals with the Python scripts, it can be tested easily.

## Usage
(examples are using zephyr-smol_llama-100m-sft-full)

### Example, `--split-max-size`
`python3 convert.py --outfile /path/to/outfile.gguf --outtype f16 /path/to/safetensors --split --split-max-size 64M`

Output: equal to what's printed to stdout from `master`, then
```
Writing the following files:
    /path/to/outfile-00001-of-00005.gguf: n_tensors = 0, total_size = negligible - metadata only
    /path/to/outfile-00002-of-00005.gguf: n_tensors = 1, total_size = 47.1M
    /path/to/outfile-00003-of-00005.gguf: n_tensors = 11, total_size = 63.6M
    /path/to/outfile-00004-of-00005.gguf: n_tensors = 32, total_size = 63.4M
    /path/to/outfile-00005-of-00005.gguf: n_tensors = 13, total_size = 19.1M

Writing shard 2/5 with 1/57 tensors remaining (of 57 total)
[1/1] Writing tensor output.weight                          | size  32128 x    768  | type F16  | T+   2

Writing shard 3/5 with 11/56 tensors remaining (of 57 total)
[ 1/11] Writing tensor token_embd.weight                      | size  32128 x    768  | type F16  | T+   2
[ 2/11] Writing tensor blk.0.attn_norm.weight                 | size    768           | type F32  | T+   3
[ 3/11] Writing tensor blk.0.ffn_down.weight                  | size    768 x   3072  | type F16  | T+   3
[ 4/11] Writing tensor blk.0.ffn_gate.weight                  | size   3072 x    768  | type F16  | T+   3
[ 5/11] Writing tensor blk.0.ffn_up.weight                    | size   3072 x    768  | type F16  | T+   3
[ 6/11] Writing tensor blk.0.ffn_norm.weight                  | size    768           | type F32  | T+   3
[ 7/11] Writing tensor blk.0.attn_k.weight                    | size    256 x    768  | type F16  | T+   3
[ 8/11] Writing tensor blk.0.attn_output.weight               | size    768 x    768  | type F16  | T+   3
[ 9/11] Writing tensor blk.0.attn_q.weight                    | size    768 x    768  | type F16  | T+   3
[10/11] Writing tensor blk.0.attn_v.weight                    | size    256 x    768  | type F16  | T+   3
[11/11] Writing tensor blk.1.attn_norm.weight                 | size    768           | type F32  | T+   3

Writing shard 4/5 with 32/45 tensors remaining (of 57 total)
[ 1/32] Writing tensor blk.1.ffn_down.weight                  | size    768 x   3072  | type F16  | T+   0
[etc...]
```

With `--split-max-size 200M` (or any number greater than the total resultant size), it gives:
```
Model has smaller size than the split threshold, not splitting

Writing the following files:
    /path/to/outfile.gguf: n_tensors = 57, total_size = 193.2M

[the rest of output is the same as in master]
```

### Example, `--split-max-tensors` with `--dry-run`, `--large-first-shard`
`python3 convert.py --outfile /path/to/outfile.gguf --outtype f16 /path/to/safetensors --split --split-max-tensors 20 --dry-run --large-first-shard`

Output: equal to what's printed to stdout from `master`, then
```
Writing the following files:
    /path/to/outfile-00001-of-00003.gguf: n_tensors = 20, total_size = 127.1M
    /path/to/outfile-00002-of-00003.gguf: n_tensors = 20, total_size = 37.5M
    /path/to/outfile-00003-of-00003.gguf: n_tensors = 17, total_size = 28.5M

Dry run, not writing files
```

With `--split-max-tensors 64` (or any number greater than the total tensor count), it gives:
```
Model has fewer tensors than the split threshold, not splitting

Writing the following files:
    /path/to/outfile.gguf: n_tensors = 57, total_size = 193.2M

Dry run, not writing files
```

**References**
- #6260
- #6688
- #6463
- #6135

christianazinn committed:
support splits in convert.py

christianazinn committed:
Support split by size and dry run to write estimated shards/filesizes

christianazinn commented:
I've added support for `--split-max-size` and `--dry-run`, taking a page out of `gguf-split.cpp`. Faced with adding split functionality to the `convert-*-to-*.py` scripts, I wonder whether this should be added to the `GGUFWriter` class itself rather than to the convert scripts, since it would be tedious to rewrite every `write_tensors` method in `convert-hf-to-gguf.py`.

The counterpoint I can see to doing this is that `GGUFWriter` should only write one *file*, since it's GG**UF**Writer and not GG**ML**Writer. It would also be very annoying to rewrite `GGUFWriter`, and I'm hesitant to touch the `gguf` package as a novice. But it's also likely nobody thought of this scenario when creating the file, so perhaps there's good reason to make these changes in the `GGUFWriter` class. @phymbert thoughts?

phymbert commented:
This is already a good start. Could you add an end to end usage in the summary?

christianazinn commented:
Sure thing (I assume you mean examples of usage and expected outputs).

I also plan to rework the implementation by consolidating code into a new `GGUFManager` class that handles multiple file writes via multiple `GGUFWriter` instances, so `GGUFWriter` still only writes to one file. This is because each `Model` in `convert-hf-to-gguf.py` has only one instance of `GGUFWriter`, so splitting would be nearly impossible there. Usage should remain the same, but the code will be fundamentally altered.

christianazinn committed:
Move split functionality to new GGUFManager class

christianazinn commented:
I'll need to implement for `convert-llama-ggml-to-gguf.py` and `convert-persimmon-to-gguf.py` soon - what are some models that require those scripts for conversion, so I can test? Also, I see `convert-lora-to-ggml.py` doesn't even use `GGUFWriter` - is that just for converting LoRA adapters? Is that something we should even add splitting for, considering the small size of LoRA adapters?

Anyway, `GGUFManager` is implemented as a near drop-in replacement for `GGUFWriter` that supports file splitting, so far only in `convert.py` (migrated from my previous commits); support for `convert-hf-to-gguf.py` is next up.

slaren commented:
`convert-llama-ggml-to-gguf.py` is for conversion of pre-gguf models. At this point it could be removed. `convert-lora-to-ggml.py` doesn't export to gguf format. `convert-persimmon-to-gguf.py` should probably be integrated into `convert-hf-to-gguf.py`, but I don't think it needs to be updated.

christianazinn commented:
Got it - will only implement for `convert-hf-to-gguf.py`. Remind me to watch memory usage while converting. Since I'm making changes to the `gguf` package, how will I push those?

slaren commented:
You can modify the gguf package in the `gguf-py` directory in this repository. There are instructions for publishing new releases in https://github.com/ggerganov/llama.cpp/blob/master/gguf-py/README.md.

christianazinn commented:
> You can modify the gguf package in the `gguf-py` directory in this repository

That's what I've been doing so far; will check out instructions to contribute, thanks!

christianazinn committed:
fix improper function signature

christianazinn commented:
Testing on Mistral 7B Instruct, this branch's `convert.py` takes up approximately the same amount of memory as that of `master`. Will need to check on larger models since the discrepancy was around 6%, 3.6G vs. 3.4G used at max. Obviously memory plays a major role in splitting larger files, which is the entire point of this PR.

christianazinn commented:
Running tests on my side for all `convert-hf-to-gguf.py` supported model architectures. What models fall under `QWenLMHeadModel` - is that just plain QWen 1?

christianazinn commented:
Will keep track of tests here as I go. Picking one model from each architecture in `convert-hf-to-gguf.py` as it exists in my branch and testing; will need assistance testing, say, vision models, which I'm not as familiar with. Also note that I went with smaller models to test the architecture; larger models should act the same, but again, tests will be needed.

It also seems like the current `convert-hf-to-gguf.py` doesn't print tensor status as it goes, which I intend to change.

- [ ] GPTNeoX: EleutherAI/gpt-neox-20b
- [ ] Bloom: bigscience/bloom-7b1
- [ ] MPT: mosaicml/mpt-7b
- [ ] Orion: OrionStarAI/Orion-14B-Chat
- [ ] Baichuan: baichuan-inc/Baichuan2-7B-Chat
- [ ] Xverse: xverse/XVERSE-7B-Chat
- [ ] Falcon: tiiuae/falcon-7b-instruct
- [ ] GPTBigCode: bigcode/gpt_bigcode-santacoder
- [ ] GPTRefact: smallcloudai/Refact-1_6B-fim
- [ ] Persimmon: adept/persimmon-8b-chat
- [ ] StableLM: stabilityai/stablelm-2-1_6b-chat
- [ ] Mistral: mistralai/Mistral-7B-Instruct-v0.2
- [ ] Llama2: meta-llama/Llama-2-7b-chat-hf
- [ ] DBRX: databricks/dbrx-instruct
- [ ] MiniCPM: openbmb/MiniCPM-V-2
- [ ] Qwen1: Qwen/Qwen-1_8B
- [ ] Qwen2: Qwen/Qwen1.5-1.8B
- [ ] Qwen MoE: Qwen/Qwen1.5-MoE-A2.7B-Chat
- [ ] GPT2: openai-community/gpt2
- [ ] Phi2: microsoft/phi-2
- [ ] Phi3: microsoft/Phi-3-mini-4k-instruct
- [ ] Plamo: pfnet/plamo-13b-instruct
- [ ] CodeShell: WisdomShell/CodeShell-7B-Chat
- [ ] InternLM: internlm/internlm2-chat-7b
- [ ] BERT: avsolatorio/GIST-Embedding-v0
- [ ] NomicBERT: nomic-ai/nomic-embed-text-v1.5
- [ ] Gemma: google/gemma-1.1-2b-it
- [ ] StarCoder2: bigcode/starcoder2-3b
- [ ] Mamba: TRI-ML/mamba-7b-rw
- [ ] Cohere: CohereForAI/c4ai-command-r-v01
- [ ] OLMo: allenai/OLMo-7B-Instruct

christianazinn commented:
Leaving a note for myself to watch merge conflicts with #6511. Development on this branch has slowed down as I'm pretty busy.

christianazinn commented:
Noting time to convert baichuan-inc/Baichuan2-7B-Chat.

New branch, `--split`, `--split-max-size 4G`:
real 6m27.788s
user 1m15.914s
sys 0m46.017s

New branch, no split:
real 7m17.661s
user 1m18.516s
sys 0m44.285s

master:
real 5m57.387s
user 1m14.567s
sys 0m48.403s

Note that these conversions were done writing the outfile over 2.5GbE, so there was considerable time spent just saving the file. Will test more later, but it doesn't seem like the change increases conversion time too significantly.

christianazinn committed:
tentative push of convert-hf-to-gguf support

mofosyne committed:
Merge branch 'master' into convert-split

mofosyne commented:
Merge attempted. Some ambiguous lines, so @christianazinn should give this a lookover to make sure the intent is still correct.

christianazinn commented:
I'll check in a few hours. But as far as merging goes, this has the potential to affect conversion workflows heavily, so I'd like to ensure we run enough tests to make sure my implementation doesn't significantly increase conversion time or memory usage or anything.

christianazinn commented:
The new `get-vocab-base-pre` functionality introduced to `convert-hf-to-gguf.py` by #6920 is throwing me off, but things look fine for the most part. Push incoming for conflict resolution; testing on Refact for `convert-hf-to-gguf.py` worked and no fundamental changes are required to `convert.py`. This will remain approximately dormant for another two weeks or so while I focus on finals, but since _the code is already almost all implemented_, if other people want to pick up and take this PR to the finish line I'd more than appreciate it.

christianazinn committed:
resolve merge + SplitArguments for easier parsing

christianazinn commented:
Finals are over and I can return to testing this branch. Will attempt a merge from master shortly, then retest each model, and if those pass without concern this branch is ready for final(ish) review.

In retrospect, this should not cause any major increases in memory or time usage over the current implementation because 1) there isn't a whole lot of added overhead and 2) Python passes everything by reference anyway as far as I'm aware.

Do people actually use temp files when converting to GGUF? That's more or less the only thing that has to be reimplemented.

I'd mostly like a review on best practices (for instance, all the filewriting logic has to go in a singular `write_to_file` method, which I can't think of a workaround for and don't imagine will be very amenable in the long run). When this branch is ready to be merged, remind me to also push to the `gguf` library.

christianazinn committed:
Merge remote-tracking branch 'origin' into convert-split

teleprint-me commented:
I've been waiting on this PR. It should be okay for me to merge it into #7379 without any issues. `convert.py` is being moved to reduce confusion in #7430.

christianazinn commented:
@teleprint-me Hold off on that for a few days - I've just found out merging `master` has introduced a major memory leak somewhere which I need to squash.

christianazinn commented:
I've narrowed the memory leak down to `gguf/lazy.py` and thereby #7075. Ironically that PR was meant to save memory (and it does in `master`). 

@compilade could I bother you to have a gander at why every time a tensor is written here, the overall RAM usage increases? It essentially just runs every tensor through an intermediate `GGUFManager` that handles splitting the tensors across files - just moving lazy tensors around should be no reason for so much RAM to accumulate (up to ten times the model size).

Frankly, this implementation is a month old at its core, so would it be better to just bin it and restart entirely?

compilade commented:
> @compilade could I bother you to have a gander at why every time a tensor is written here, the overall RAM usage increases?

@christianazinn Thanks for letting me know about this. Before I dig deeper into it, my hypothesis is that something holds onto references to tensors that were made "eager".

If there's a list of all tensors, even if they are lazy at first, when they are written to a file (one of the things that can make them eager), the old reference to the LazyNumpyTensors now also point to the calculated data in their `_data` field. This was done to avoid having to re-calculate stuff when one lazy tensor is re-used in many places, but it assumes references are not kept longer than they should, which might not be true with `GGUFManager`.

christianazinn commented:
> Before I dig deeper into it, my hypothesis is that something holds onto references to tensors that were made "eager".

Thanks @compilade and you seem to be right - after making sure every reference is deleted when no longer used, the RAM usage normalizes. Pushing a fix that, among others, adds an explicit `del tensor` to `gguf_writer`'s `write_tensors_to_file` - should have no adverse side effects because under normal circumstances the garbage collector deals with it anyway. Back to testing!

Also removing all changes to `convert.py` in light of #7430.

christianazinn committed:
Fix eager tensor memory leak and remove convert.py changes

Removed a memory leak caused by unexpected reference retention to eager tensors.

Also removed GGUFManager functionality in convert.py in favor of specializing for convert-hf-to-gguf.py.

christianazinn committed:
refactor SplitStrategy to be a deque

Instead of having SplitStrategy have a `data` field that is a deque, just have SplitStrategy be a subclass of deque itself.

christianazinn commented:
Ran tests (have a big Excel spreadsheet if anyone wants it) and there don't appear to be any major differences in conversion time or memory usage, as expected. I just need to run tests on tempfile and outtype Q8_0 (because for some reason that doesn't play nice) and this PR will be ready for review.

christianazinn committed:
fix Q8 quantization

christianazinn committed:
remove unnecessary imports in gguf_manager

christianazinn commented:
Fixed an error with Q8 quantization. Tempfile appears to be working as well. Going to merge master one last time before marking as ready for review.

christianazinn committed:
Merge branch 'master' into convert-split

christianazinn committed:
fix final? merge issue

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -74,9 +77,13 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
             self.part_names = Model.get_model_part_names(self.dir_model, ".bin")
+            
         self.hparams = Model.load_hparams(self.dir_model)
+        self.gguf_writer = gguf.GGUFManager(fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], split_arguments,
+                                            endianess=self.endianess, use_temp_file=self.use_temp_file)
```
This needs to be lower in `__init__`, or else `fname_out` won't always be correct, especially with outtype templating.

christianazinn replied:
Fixed, thank you. The placement caught my eye in the diffs but I didn't realize it was important.

compilade commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,543 @@
+        return f"{num:.1f}T - over 1TB, --split recommended"
+
+
+# ideally this has most of the same signatures as GGUFWriter so it's nearly a drop-in replacement
```
It's likely possible to duplicate even less of the methods by doing something similar to the Lazy tensors from `gguf-py/gguf/lazy.py`. Something with `__getattr__` to intercept missing attributes and redirect them.

But it might make `kv_data` management more complicated, though.

christianazinn replied:
Not familiar with how those work (other than they gave me a memory leak, haha) but I'll take a look. Another thought on reducing boilerplate - could we have `GGUFManager` subclass `GGUFWriter` somehow?

christianazinn committed:
fix gguf_writer placement and remove comments

christianazinn committed:
oops, actually fix gguf_writer placement

christianazinn commented:
By the way, this shouldn't be marked "examples." Would appreciate a full review or so soon(tm).

ggerganov commented:
I've just had a look at the implementation and it seems to add too much complexity for a feature, that I would guess is needed very rarely. The `GGUFManager` seems like over-engineering, so I recommend either significantly simplifying the implementation or drop this feature completely

christianazinn commented:
> I've just had a look at the implementation and it seems to add too much complexity for a feature, that I would guess is needed very rarely. The `GGUFManager` seems like over-engineering, so I recommend either significantly simplifying the implementation or drop this feature completely

Thanks for the feedback. Is it the boilerplate, the implementation in write_to_file, the unnecessity of having an extra class that's basically just a tweaked GGUFWriter, or something else that seems overengineered? The split_strategy method is used by the C++ script and there doesn't seem to be a better way to set up shard data.

christianazinn committed:
reduce duplicated code from gguf_writer

christianazinn committed:
further simplify GGUFManager

christianazinn committed:
simplify even further and standardize with GGUFWriter

christianazinn committed:
reduce diffs with master

christianazinn commented:
@ggerganov I should've been able to considerably reduce complexity (not sure why I didn't think of subclassing `GGUFWriter` before) so the functionality is as compact as possible. Could you look it over again? If it's still too complex, could you explain what exactly about it is? It should be simpler than `examples/gguf-split` now.

We can also have `convert-hf-to-gguf.py` fall back to base `GGUFWriter` for non-split operations so this PR only affects split conversions, if desired.

christianazinn committed:
form shards while adding tensors, SHA256 sums agree with master

christianazinn commented:
Went ahead and made the change to remove `SplitStrategy` because I thought it wasn't worth the trouble keeping it. The logic looks a lot nicer now, and in retrospect the original wasn't "simple" in the slightest. Thanks again for getting me to fix it @ggerganov. SHA256 sums for vocab and full models are identical to master, as are split sums with splits generated by `examples/gguf-split` with both small and large first shard. Hoping this is now satisfactory.

compilade commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+
+LLM_KV_SPLIT_NO = "split.no"
+LLM_KV_SPLIT_COUNT = "split.count"
+LLM_KV_SPLIT_TENSORS_COUNT = "split.tensors.count"
```
I think these three should go in `gguf-py/gguf/constants.py`

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -57,7 +57,7 @@ class GGUFWriter:
     def __init__(
         self, path: os.PathLike[str] | str, arch: str, use_temp_file: bool = True,
-        endianess: GGUFEndian = GGUFEndian.LITTLE,
+        endianess: GGUFEndian = GGUFEndian.LITTLE, dont_add_architecture: bool = False
```
Maybe rename `dont_add_architecture` to `add_architecture` with a default to `True` instead? Otherwise it seems this is always used with `not`, both when set and when used.

And the double negation can be somewhat confusing.

christianazinn replied:
I was thinking about that. Honestly I have no idea why I went with the negative anymore.

christianazinn replied:
Resolved with 3328b0a.

compilade commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+
+        self.state = WriterState.KV_DATA
+
+    def write_tensors_to_file(self, progress: bool = False) -> None:
```
```suggestion
    def write_tensors_to_file(self, *, progress: bool = False) -> None:
```

This is otherwise incompatible with `GGUFWriter`. It's only accepted as a keyword argument, not also positional.

compilade commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+    def format_n_bytes_to_str(num: int) -> str:
+        if num == METADATA_ONLY_INDICATOR:
+            return "negligible - metadata only"
+        num = float(num)
```
The `num` variable was declared to be an `int`, so a `float` should not be assigned to it. Use another variable for this.

christianazinn replied:
Resolved with 1cbab22.

christianazinn commented on convert-hf-to-gguf.py:
```diff
@@ -23,7 +23,9 @@
 if 'NO_LOCAL_GGUF' not in os.environ:
     sys.path.insert(1, str(Path(__file__).parent / 'gguf-py'))
-import gguf
+import importlib
+gguf = importlib.import_module("gguf-py.gguf")
+# import gguf
```
This is a temporary fix to load the updated `gguf` package during development, because importing from local path complains about the dash in the directory name. When this PR is merged (or when I get the greenlight on the `gguf` package part), I will push the changes to the `gguf` package and replace these lines with `import gguf` again.

compilade replied:
> This is a temporary fix to load the updated `gguf` package during development, because importing from local path complains about the dash in the directory name.

But I think the `NO_LOCAL_GGUF` path insertion thing should already do this, no? If you happen to have the `NO_LOCAL_GGUF` environment variable defined, maybe try to `unset NO_LOCAL_GGUF`?

> When this PR is merged (or when I get the greenlight on the gguf package part), I will push the changes to the gguf package and replace these lines with import gguf again.

The source of the `gguf` package ***is*** `gguf-py` in the `llama.cpp` repo, from what I understand. So you don't need to separately push changes to "the `gguf` package", since you're already modifying it.

What could be done however, to eventually update the package in pypi, would be to bump the version and tag it as described in <https://github.com/ggerganov/llama.cpp/tree/master/gguf-py#automatic-publishing-with-ci>. But this can also be done separately.

christianazinn replied:
Oh. I had meant pypi updating, but I didn't realize it was taken straight from the local `gguf-py` directory. Strange, I'll look into this when I get home.

christianazinn replied:
Confirmed that it does seem like a simple `import gguf` does the trick. Fixed in c6ae1d6.

christianazinn commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+
+KVTempData: TypeAlias = dict[str, tuple[Any, GGUFValueType]] # {key: (value, type)}
+TensorTempData: TypeAlias = tuple[str, np.ndarray[Any, Any], GGMLQuantizationType] # (tensor name, tensor data, tensor dtype)
+Shard: TypeAlias = list[os.PathLike[str], int, int, deque[TensorTempData]] # [shard filename, shard tensor count, shard size, [tensor data]]
```
Good idea. Yeah, I had tried a tuple, then realized immutability wouldn't work. Dataclass it is.

christianazinn commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+
+class GGUFManager(GGUFWriter):
+    kv_data: KVTempData
+    tensors: list[TensorTempData]
```
Let me know what you decide. For now it's not too difficult to just rename the field, since we don't actually care about the superclass `tensors` field - we don't even call the superclass constructor, as we only want the signatures of most of the functions.

christianazinn replied:
It turns out we don't even use this field anymore, so I deleted it.

christianazinn commented on gguf-py/gguf/gguf_manager.py:
```diff
@@ -0,0 +1,254 @@
+            self.shards[0][0] = self.path
+        else:
+            for i in range(len(self.shards)):
+                self.shards[i][0] = self.path.with_name(SHARD_NAME_FORMAT.format(self.path.stem, i + 1, len(self.shards)))
```
Thanks. It should be made explicit.

christianazinn replied:
The `self.path` field is never type annotated so it should be fine to just initialize `self.path` as a `Path` object (`Path(path)`) in the `GGUFManager` constructor. Resolved with 83e4a3f.

christianazinn committed:
re-add type hint

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
GGUFWriter compatibility fix

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Shard dataclass and un-negative dont_add_architecture

christianazinn committed:
type consistency in format_n_bytes_to_str

christianazinn committed:
move kv keys to constants.py

christianazinn committed:
make pathlib explicit

christianazinn commented:
So the `tqdm` progress bar in `GGUFWriter` uses 1000-based byte sizes (KB, MB, GB), while `GGUFManager` at time of writing uses 1024-based (KiB, MiB, GiB). I'm changing to 1000-based because HuggingFace recommends 5GB splits (which you see with most safetensors/pytorch models), but the `examples/gguf-split` script appears to use 1024-based. So I might open another PR to change that.

christianazinn committed:
base-1024 bytes to base-1000

ggerganov commented:
Would it make sense to rename `GGUFManager` to `GGUFWriterSplit`? Might also consider merging the split logic entirely in `GGUFWriter` and avoid introducing the extra class - though feel free to ignore this suggestion if you don't think it would make sense. We can merge when @compilade approves

christianazinn committed:
rename GGUFManager to GGUFWriterSplit

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -93,7 +96,8 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
         self.fname_out = fname_out.parent / fname_out.name.format(ftype_lw, outtype=ftype_lw, ftype=ftype_lw, OUTTYPE=ftype_up, FTYPE=ftype_up)
-        self.gguf_writer = gguf.GGUFWriter(self.fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], endianess=self.endianess, use_temp_file=self.use_temp_file)
+        self.gguf_writer = gguf.GGUFWriterSplit(fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], split_arguments,
+                                            endianess=self.endianess, use_temp_file=self.use_temp_file)
```
There are quite a lot of complaints from the `flake8` linter.

See if you can fix this to make the tests pass: <https://github.com/ggerganov/llama.cpp/actions/runs/9417611088/job/25943176592?pr=6942>

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -323,12 +327,16 @@ def write_tensors(self):
 
     def write(self):
         self.write_tensors()
+        self.gguf_writer.init_shards()
```
The type of `self.gguf_writer` is `GGUFWriter` which does not have the `init_shards` method.

christianazinn replied:
Right, will fix that type annotation. Do we then want to rename `self.gguf_writer` itself or can we leave it as such?

compilade replied:
Leaving it as `self.gguf_writer` should be fine.

christianazinn replied:
Thanks. Fixed in c6ae1d6.

christianazinn committed:
Update gguf-py/gguf/constants.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
fix convert-hf-to-gguf.py permissions

compilade commented:
@christianazinn Warning, <https://github.com/ggerganov/llama.cpp/pull/6942/commits/5f29d4a6172948b130f5f9b8fab3cf7ee9778425> changed all lines in `convert-hf-to-gguf.py` to CRLF line endings

christianazinn commented:
> @christianazinn Warning, [5f29d4a](https://github.com/ggerganov/llama.cpp/commit/5f29d4a6172948b130f5f9b8fab3cf7ee9778425) changed all lines in `convert-hf-to-gguf.py` to CRLF line endings

Gah, you're joking. And it didn't fix anything either. Give me a minute

christianazinn commented on convert-hf-to-gguf.py:
Strange, that didn't do anything. It appears the same has happened for `examples/convert-legacy-llama.py`. Both appear as `-rwxrwxrwx` on this development environment.

compilade replied:
Oh, right, your on a Windows filesystem. Nevermind then. This is easier to fix on a filesystem which supports Unix-style permissions.

christianazinn replied:
Haha, yeah, the CRLF fiasco gave it away. Was trying to do this via WSL2 but I guess it inherits Windows file permissions through DrvFs.

compilade replied:
It's alright. I've pushed a fix for this in <https://github.com/ggerganov/llama.cpp/pull/6942/commits/e093dfba9f2ba2b8f5345c747c8e96f7776789ca>. If you're not happy with it, feel free to force push it away.

christianazinn replied:
Thanks. It's also happened to `examples/convert-legacy-llama.py` but I think `git clone`ing directly to the WSL filesystem should let me fix that.

christianazinn committed:
fix line endings

christianazinn committed:
Update gguf-py/gguf/gguf_writer_split.py

Co-authored-by: compilade <git@compilade.net>

compilade committed:
convert-hf : restore executable file permission

christianazinn committed:
examples/convert-legacy-llama.py: restore executable file permission

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -299,6 +300,7 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
                     tensor.tofile(self.fout)
                     bar.update(tensor.nbytes)
                     self.write_padding(self.fout, tensor.nbytes)
+                    del tensor
```
Looking back on this, yeah, that makes no sense. Considering `tensor` is `pop`ped from the list, there shouldn't be a reference left to it if everything else works as intended, and there isn't - commenting out this line shows no change in memory usage.

I think I was conflating cause and effect of a few different things when I made this fix. But since there's no harm in leaving it, it'll stay. Thanks for pointing that out.

christianazinn committed:
reinstate original gguf package import and fix type annotation

christianazinn committed:
attempt to appease the linter

christianazinn committed:
attempt 2 to appease the linter

christianazinn committed:
attempt 3 to appease the linter

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2793,6 +2801,26 @@ def parse_args() -> argparse.Namespace:
     )
+    parser.add_argument(
+        "--split", action="store_true",
+        help="split the converted model into multiple files"
+    )
+    parser.add_argument(
+        "--split-max-tensors", type=int,
+        help="max tensors in each split"
+    )
+    parser.add_argument(
+        "--split-max-size", type=str,
+        help="max size per split N(M|G)"
+    )
+    parser.add_argument(
+        "--dry-run", action="store_true",
+        help="only print out a split plan and exit, without writing any new files"
+    )
+    parser.add_argument(
+        "--small-first-shard", action="store_true",
+        help="do not add tensors to the first shard (disabled by default)"
```
The other arguments above the new ones all have a trailing comma after the `help="..."`.

christianazinn committed:
comma consistency

christianazinn commented:
@compilade Since you already have #7827 approved, why don't we do this:
We can merge that first and I'll add the simplification into this PR.
Then I'll attempt to merge `GGUFSplitWriter` functionality into `GGUFWriter` just to see how clean it can actually be. If you like it, we can keep it; if not, we can revert the commit. I see your point that we could have `GGUFWriter` write a *model* rather than a *file* and think we could at least try it.
After that we can finally merge this. Let me know what you think.

compilade commented:
> Then I'll attempt to merge `GGUFSplitWriter` functionality into `GGUFWriter` just to see how clean it can actually be.

I think `self.fout`, `self.tensors` and `self.kv_data` from `GGUFWriter` could actually be made into lists, one element (file/dict) per file.

> After that we can finally merge this. Let me know what you think.

Yes, I think this is a good plan. It will make this simpler, and you will have more control over how exactly splitting is simplified.

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -93,7 +94,8 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
         # allow templating the file name with the output ftype, useful with the "auto" ftype
         self.fname_out = fname_out.parent / fname_out.name.format(ftype_lw, outtype=ftype_lw, ftype=ftype_lw, OUTTYPE=ftype_up, FTYPE=ftype_up)
-        self.gguf_writer = gguf.GGUFWriter(self.fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], endianess=self.endianess, use_temp_file=self.use_temp_file)
+        self.gguf_writer = gguf.GGUFWriterSplit(fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], split_arguments,
```
```suggestion
        self.gguf_writer = gguf.GGUFWriterSplit(self.fname_out, gguf.MODEL_ARCH_NAMES[self.model_arch], split_arguments,
```
Otherwise the name template is not applied

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2793,6 +2799,26 @@ def parse_args() -> argparse.Namespace:
         "--verbose", action="store_true",
         help="increase output verbosity",
     )
+    parser.add_argument(
+        "--split", action="store_true",
+        help="split the converted model into multiple files",
```
Shouldn't this be implied by `--split-max-size` and/or `--split-max-tensors`? Otherwise nothing special happens when only using `--split-max-size`, and this can be confusing, especially since `--split` alone can't be used.

christianazinn replied:
Yes, that happened to me a few times in testing actually. Give me some time to get back to my dev environment.

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -2793,6 +2799,26 @@ def parse_args() -> argparse.Namespace:
+        help="only print out a split plan and exit, without writing any new files",
+    )
+    parser.add_argument(
+        "--small-first-shard", action="store_true",
+        help="do not add tensors to the first shard (disabled by default)",
```
`shard` vs `split`, might not be obvious that this is related to splits. Should the other options be renamed to `--shard-max-size` and `--shard-max-tensors`?

(You can answer "no" too)

christianazinn replied:
Renaming for consistency with examples/gguf-split is probably the way to go here.

compilade commented on convert-hf-to-gguf.py:
There's this line in `main()`:

```python3
logger.info(f"Exporting model to '{model_instance.fname_out}'")
```

and it prints a name which isn't used when split options are given.

christianazinn replied:
I think the same happens at the end when it prints "Model successfully exported to (so-and-so)". We can handle this fine because we know the files after we've written them (obviously). But we don't know how many splits we need before we've started writing tensors, which means the example you bring up would be a problem. Perhaps we can be generic with "(fname)-0000*-of-0000*.gguf" or something?

christianazinn replied:
Because we actually print the files we're writing to in `print_plan` (which is always called by `write_header_to_file`), we don't need these lines anymore. I've replaced them with generic "Exporting model..." etc. lines.

compilade reviewed (commented):
It does work! (although `./gguf-split`'s handling of file suffixes is worse. At least this Python implementation is better in this aspect)

I've tested with a `q8_0` quant of <https://huggingface.co/jtatman/TinyMistral-248m-v2.5-4x-Moe>.

```console
$ ls -sh tinymistral-moe-convert-hf-100M/
total 712M
85M tinymistral-moe-q8_0-00001-of-00008.gguf  91M tinymistral-moe-q8_0-00005-of-00008.gguf
91M tinymistral-moe-q8_0-00002-of-00008.gguf  88M tinymistral-moe-q8_0-00006-of-00008.gguf
88M tinymistral-moe-q8_0-00003-of-00008.gguf  91M tinymistral-moe-q8_0-00007-of-00008.gguf
91M tinymistral-moe-q8_0-00004-of-00008.gguf  91M tinymistral-moe-q8_0-00008-of-00008.gguf
$ sha256sum tinymistral-moe-convert-hf-100M/*
85d7530f12fd3918c0365b23250a86187810d8da1e5667c740b78ef0994f97a6  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00001-of-00008.gguf
f391c85dab1e1d3b870c3c49b78fcd4111979e30d2d759df0ed494acdfa12749  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00002-of-00008.gguf
595e07a71d09f525a298999b59ac122179840557bd4770ccfdfa7acb33913e7a  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00003-of-00008.gguf
44984171e33250c9ed6b9e7c3b7d0798edf6148f25b5f75700c36b15fc36f5a2  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00004-of-00008.gguf
6c8c41f3e9433a42015a6a63637f13531613ca9f1b18e9774e57726604ff9b15  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00005-of-00008.gguf
df4bb369215013afc1bc0a9f70e79cc13ae3405f7c6b9a3b63d964852d28f347  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00006-of-00008.gguf
1f48a1caf89f3d11e9cddf36c33f7a75bff2a8d9da7e9ee4fb17b3c83cdc26a2  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00007-of-00008.gguf
fe66ff1cade4da482f99780d7ccf419138e848f8b3ce6a7ef47f09c9f592e3c3  tinymistral-moe-convert-hf-100M/tinymistral-moe-q8_0-00008-of-00008.gguf
$ sha256sum tinymistral-moe-gguf-split-100M/*
85d7530f12fd3918c0365b23250a86187810d8da1e5667c740b78ef0994f97a6  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00001-of-00008.gguf
f391c85dab1e1d3b870c3c49b78fcd4111979e30d2d759df0ed494acdfa12749  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00002-of-00008.gguf
595e07a71d09f525a298999b59ac122179840557bd4770ccfdfa7acb33913e7a  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00003-of-00008.gguf
44984171e33250c9ed6b9e7c3b7d0798edf6148f25b5f75700c36b15fc36f5a2  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00004-of-00008.gguf
6c8c41f3e9433a42015a6a63637f13531613ca9f1b18e9774e57726604ff9b15  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00005-of-00008.gguf
df4bb369215013afc1bc0a9f70e79cc13ae3405f7c6b9a3b63d964852d28f347  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00006-of-00008.gguf
1f48a1caf89f3d11e9cddf36c33f7a75bff2a8d9da7e9ee4fb17b3c83cdc26a2  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00007-of-00008.gguf
fe66ff1cade4da482f99780d7ccf419138e848f8b3ce6a7ef47f09c9f592e3c3  tinymistral-moe-gguf-split-100M/tinimistral-moe-q8_0.gguf-00008-of-00008.gguf
```

christianazinn committed:
Update convert-hf-to-gguf.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
edit cmd line args

christianazinn committed:
Merge branch 'master' into convert-split

christianazinn committed:
use simplification from #7827

christianazinn commented:
03cc9bc should be the last commit featuring `GGUFSplitWriter` - will now try integrating it into `GGUFWriter`.

christianazinn committed:
kv/ti data are still wrong

christianazinn commented:
Pushing because I'm going to bed, will fix what's wrong in the morning (including editorconfig and linter). Files are getting the right tensors but not the right metadata: first shard has first line `GGUF]eneral.architecturllama` when displayed via `head` but main has it as `GGUFfgeneral.architecturllama`. Second shard is apparently getting the first shard tensor metadata. Confusing.

compilade commented:
> Second shard is apparently getting the first shard tensor metadata. Confusing.

There's no need to separately pack the split metadata. It can be directly included in the `kv_data` dict (less confusing).

Here, `self.kv_data` is a `list[dict[str, GGUFValue]]`, and `self.tensors` is a `list[dict[str, TensorInfo]]`. Oh, and `self.fout` is a `list[BufferedWriter] | None` in this example.

I did not test this yet, but it should give an idea of a different way to approach the inclusion of shard metadata.

```python3
    def write_header_to_file(self, path: Path | None = None) -> None:
        self.open_output_file(path)

        if self.state is not WriterState.EMPTY:
            raise ValueError(f'Expected output file to be empty, got {self.state}')
        assert self.fout is not None
        assert len(self.fout) == len(self.tensors)
        assert len(self.kv_data) == 1

        shards: int = len(self.fout)
        total_tensors: int = sum(len(t) for t in self.tensors)

        for i, fout in enumerate(self.fout):
            if shards > 1:
                shard_info = {
                    Keys.Split.LLM_KV_SPLIT_NO: GGUFValue(i, GGUFValueType.UINT16),
                    Keys.Split.LLM_KV_SPLIT_COUNT: GGUFValue(shards, GGUFValueType.UINT16),
                    Keys.Split.LLM_KV_SPLIT_TENSORS_COUNT: GGUFValue(total_tensors, GGUFValueType.INT32),
                }
                if i == 0:
                    self.kv_data[0] = {**self.kv_data[0], **shard_info}
                else:
                    self.kv_data.append(shard_info)

            fout.write(self._pack("<I", GGUF_MAGIC, skip_pack_prefix = True))
            fout.write(self._pack("I", GGUF_VERSION))
            fout.write(self._pack("Q", len(self.tensors[i])))
            fout.write(self._pack("Q", len(self.kv_data[i])))

        self.flush()
        self.state = WriterState.HEADER
```

And then when `self.kv_data` is actually written to the files, it can stay very simple:

```python3
    def write_kv_data_to_file(self) -> None:
        if self.state is not WriterState.HEADER:
            raise ValueError(f'Expected output file to contain the header, got {self.state}')
        assert self.fout is not None
        assert len(self.fout) == len(self.kv_data)

        for fout, kv_data in zip(self.fout, self.kv_data):
            kv_bytes = bytearray()

            for key, val in kv_data.items():
                kv_bytes += self._pack_val(key, GGUFValueType.STRING, add_vtype=False)
                kv_bytes += self._pack_val(val.value, val.type, add_vtype=True)

            fout.write(kv_bytes)

        self.flush()
        self.state = WriterState.KV_DATA
```
Feel free to take whatever you want from these snippets.

> I'm going to bed

Yeah, it's late for me too. Good night :)

christianazinn committed:
try to refactor kv data (still fails)

christianazinn committed:
fix ti data messiness

christianazinn committed:
Merge branch 'master' into convert-split

christianazinn commented:
Apparently I was just reusing a loop variable in writing `ti_data` and that's what caused the difference in file hashes. They now match `master` (and the previous version) on all cases.

christianazinn committed:
tidy up

christianazinn committed:
fix linting

christianazinn committed:
actually make the linter happy

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -608,6 +735,46 @@ def _pack_val(self, val: Any, vtype: GGUFValueType, add_vtype: bool) -> bytes:
+        fout.write(self._pack(fmt, value, skip_pack_prefix))
+
+    @staticmethod
+    def get_tensor_size(tensor) -> int:
+        try:
+            return tensor.data_type.elements_to_bytes(np.prod(tensor.shape))
+        except AttributeError: # numpy ndarray[Any, Any]
+            return tensor.nbytes
```
With `TensorInfo`, the size in bytes is always available with the `nbytes` field, so this function doesn't seem necessary.

compilade reviewed (commented):
Hopefully this isn't too many comments :)

Some of them are related to things a static type checker could point out. (I get 25 type errors (from `pyright`) in `gguf-py/gguf/gguf_writer.py` in <https://github.com/ggerganov/llama.cpp/pull/6942/commits/49b9fbe9420365cbb6f7a8f312a41883a98ea186>, while there are none on `master`)

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -247,7 +353,18 @@ def add_tensor_info(
 
-        self.tensors[name] = TensorInfo(shape=tensor_shape, dtype=dtype, nbytes=tensor_nbytes)
+        # create splits as necessary, such as to start it off
+        if (len(self.tensors) == self.split_arguments.small_first_shard \
+            # or split when over tensor limit
+            or (self.split_arguments.split_style == SplitStyle.TENSORS \
+                and len(self.tensors[-1]) >= self.split_arguments.split_max_tensors) \
+            # or split when over size limit
+            or (self.split_arguments.split_style == SplitStyle.SIZE \
+                and GGUFWriter.get_tensors_total_size(self.tensors[-1].values()) + tensor_nbytes > self.split_arguments.split_max_size)):
```
Interesting thought. `examples/gguf-split` doesn't support this so I hadn't thought of that. But this shouldn't be too hard to implement.

christianazinn replied:
It does work, though I'm not sure how useful it is (for that matter I don't know of any use cases for `--split-max-tensors` in the first place).

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -608,6 +736,46 @@ def _pack_val(self, val: Any, vtype: GGUFValueType, add_vtype: bool) -> bytes:
+            return tensor.nbytes
+        
+    @staticmethod
+    def get_tensors_total_size(tensors) -> int:
+        return sum(GGUFWriter.get_tensor_size(ti) for ti in tensors)
```
Definitely; am removing it. `get_tensor_size` itself is a holdover from when I think we were using two different datatypes for `tensor`.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -54,12 +84,18 @@ class WriterState(Enum):
     temp_file: tempfile.SpooledTemporaryFile[bytes] | None
-    tensors: dict[str, TensorInfo]
-    kv_data: dict[str, GGUFValue]
+    tensors: list[dict[str, TensorInfo | np.ndarray[Any, Any]]]
```
Oh. Oops.

christianazinn committed:
cleanup round 1

christianazinn committed:
remove SplitStrategy, SplitArguments

christianazinn committed:
appease linter

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -45,6 +57,25 @@ class GGUFValue:
+            else SplitStyle.SIZE if self.split_max_size \
+            else SplitStyle.NONE
+        self.dry_run = args.dry_run
+        self.small_first_shard = args.no_tensor_first_split
```
It's not pretty, but it works.

christianazinn commented:
> Some of them are related to things a static type checker could point out.

That seems to have been a running theme. I really ought to get one. It should be much better now.

christianazinn commented on convert-hf-to-gguf.py:
```diff
@@ -65,7 +65,8 @@ class Model:
 
-    def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path, is_big_endian: bool, use_temp_file: bool, eager: bool, model_name: str | None):
+    def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path, is_big_endian: bool, use_temp_file: bool, eager: bool,
+                 model_name: str | None, split_max_tensors: int = 0, split_max_size: int = 0, dry_run: bool = 0, small_first_shard: bool = 0):
```
Can't believe I missed that.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -106,36 +141,72 @@ def open_output_file(self, path: os.PathLike[str] | str | None = None) -> None:
+
     def write_header_to_file(self, path: os.PathLike[str] | str | None = None) -> None:
+        self.verify_arguments()
+        self.print_plan(path)
```
`self.path` really should be `Path | None`: `fname_out` is a `Path` in `convert-hf-to-gguf.py` so there's no reason for it to be anything else (other than `None`) in `GGUFWriter`.

christianazinn replied:
This also makes `path` pointless in the signature of `open_output_file` (unless we want to duplicate the code). Is removing it acceptable?

compilade replied:
> `self.path` really should be `Path | None`

True. I agree for `self.path`. But I think the `path` argument of `GGUFWriter.__init__` should stay a `Pathlike[str] | str | None`, because some scripts like `gguf-py/examples/writer.py` and `examples/llava/convert-image-encoder-to-gguf.py` use a `str` for the path.

https://github.com/ggerganov/llama.cpp/blob/e95beeb1fc4621826ddd616776dbdf717366bf5c/examples/llava/convert-image-encoder-to-gguf.py#L167-L168

https://github.com/ggerganov/llama.cpp/blob/e95beeb1fc4621826ddd616776dbdf717366bf5c/gguf-py/examples/writer.py#L16

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -247,7 +321,18 @@ def add_tensor_info(
 
-        self.tensors[name] = TensorInfo(shape=tensor_shape, dtype=dtype, nbytes=tensor_nbytes)
+        # create splits as necessary, such as to start it off
+        if (len(self.tensors) == self.small_first_shard \
```
This is shorthand for:
```python
if (self.small_first_shard and len(self.tensors) == 1 \
    or (not self.small_first_shard and len(self.tensors) == 0)
```
but with your other recommended changes I think we can make this simpler.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
It looks like the last one deals with `temp_file` which there doesn't seem to be a good way to handle splitting for. For now I'm just going to prevent using temp file with splits.

compilade replied:
> For now I'm just going to prevent using temp file with splits.

I'm fine with that.

It might still be possible to allow `temp_file` by adding one tempfile per shard, though this might use too much memory, since the temp files are spooled in memory (256MiB by default).

https://github.com/ggerganov/llama.cpp/blob/0417104397e54dce71cd2c6e9c23b11f2acf0d60/gguf-py/gguf/gguf_writer.py#L320

So I think preventing temp files when splitting like you did is the right choice.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -76,25 +85,51 @@ class GGUFWriter:
         ))
         self.state = WriterState.NO_FILE
 
+        if self.small_first_shard:
+            self.tensors.append(dict())
+
```
Best to do this the other way: remove the check in `add_tensor_info` and put it here.

christianazinn committed:
fix typing and clean up

christianazinn committed:
fix linting

christianazinn commented:
Finally bothered to configure flake8 and Pylance on my local machine so hopefully I don't keep messing up linting/typing in the future. Though it's probably still failing tests somewhere, somehow...

compilade reviewed (commented):
This is getting good! Zero type errors from `pyright`. Now my comments focus more on the code's behavior.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -247,7 +301,14 @@ def add_tensor_info(
 
-        self.tensors[name] = TensorInfo(shape=tensor_shape, dtype=dtype, nbytes=tensor_nbytes)
+        # split when over tensor limit
+        if (self.split_max_tensors != 0 and len(self.tensors[-1]) >= self.split_max_tensors \
+                # or split when over size limit
+                or self.split_max_size != 0 and sum(ti.nbytes for ti in self.tensors[-1].values()) + tensor_nbytes > self.split_max_size):
```
We can do that. Because we write the tensor to the shard immediately after, this only affects the very first tensor... but because `token_embd.weight` is often the first and one of the biggest tensors, that's one of the most common cases.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -106,36 +122,76 @@ def open_output_file(self, path: os.PathLike[str] | str | None = None) -> None:
+        assert self.fout is not None
+        total_splits = len(self.fout)
+        for i in range(total_splits):
+            # just see whether it exists
+            try:
+                self.kv_data[i]
+            except IndexError:
+                self.kv_data.append(dict())
```
Yes, this works, and is much less hacky.

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -293,44 +356,49 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
+                    total_bytes = sum(ti.nbytes for ti in tensors.values())
 
-                bar = tqdm(desc="Writing", total=total_bytes, unit="byte", unit_scale=True)
+                    bar = tqdm(desc="Writing", total=total_bytes, unit="byte", unit_scale=True)
```
Good idea. Since some people might like to see progress within the shard, I've also added a `tqdm` progress bar underneath for the individual shard.

christianazinn committed:
progress bar, fix split logic

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -247,7 +297,16 @@ def add_tensor_info(
 
-        self.tensors[name] = TensorInfo(shape=tensor_shape, dtype=dtype, nbytes=tensor_nbytes)
+        # make sure there is at least one tensor before splitting
+        if (len(self.tensors[-1]) > 0
+                # split when over tensor limit
+                and (self.split_max_tensors != 0 and len(self.tensors[-1]) >= self.split_max_tensors)
+                # or split when over size limit
+                or (self.split_max_size != 0 and sum(ti.nbytes for ti in self.tensors[-1].values()) + tensor_nbytes > self.split_max_size)):
+
+            self.tensors.append(dict())
```
Thanks. Also forgot to change `dict()` initializer to `{}`, so this fixes that.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
catch oversights

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -293,43 +358,60 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
-                self.write_padding(self.fout, ti.nbytes)
-                ti.tensor = None
+                if len(self.fout) > 1:
+                    shard_bar = tqdm(desc=f"Shard (1/{len(self.fout)})", total=total_bytes, unit="byte", unit_scale=True)
```
To distinguish this placeholder from the actual first shard
```suggestion
                    shard_bar = tqdm(desc=f"Shard (0/{len(self.fout)})", total=None, unit="byte", unit_scale=True)
```

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -293,43 +358,60 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
+                    shard_bar.set_description(f"Shard ({i + 1}/{len(self.fout)})")
+                    total = sum(ti.nbytes for ti in tensors.values())
+                    # bar behaves weirdly when total is 0
+                    if total > 0:
+                        shard_bar.reset(total=total)
```
I've tested this, and `total` can be set to `None` to avoid the weirdness at `0`.
```suggestion
                    shard_bar.reset(total=(total if total > 0 else None))
```

christianazinn replied:
Nice catch. Weird how that works.

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -293,43 +358,60 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
-                if bar is not None:
-                    bar.update(ti.nbytes)
-                self.write_padding(self.fout, ti.nbytes)
-                ti.tensor = None
+                if len(self.fout) > 1:
+                    shard_bar = tqdm(desc=f"Shard (1/{len(self.fout)})", total=total_bytes, unit="byte", unit_scale=True)
```
I don't know about you, but from trying it a few times, I feel like the global progress `bar` should be on the bottom, while the `shard_bar` should be right above it. This simply requires swapping their order here (and also maybe in the other places for consistency).

This is subjective, but I like to have the slower-moving `bar` anchored on the bottom, and not sandwiched between the previous logs and a fast-moving `shard_bar`.

christianazinn replied:
We can do that. Personally I prefer it the other way but I see how you'd want the things that complete first on the top.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
swap bar orders

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -282,9 +345,11 @@ def write_tensor_data(self, tensor: np.ndarray[Any, Any]) -> None:
-        tensor.tofile(self.fout)
-        self.write_padding(self.fout, tensor.nbytes)
+
+        for fout in self.fout:
+            self.write_padding(fout, fout.tell())
+            tensor.tofile(fout)
+            self.write_padding(fout, tensor.nbytes)
```
This function (`write_tensor_data`) is only used by `convert-legacy-llama.py` and `gguf-new-metadata.py` which don't yet use splitting options, but if they ever do, I don't think the tensor should be written to *all* files.

<details><summary>Maybe something like this could work? (click to expand)</summary>

```diff
diff --git a/gguf-py/gguf/gguf_writer.py b/gguf-py/gguf/gguf_writer.py
index 8042ecee..9f07b4b2 100644
--- a/gguf-py/gguf/gguf_writer.py
+++ b/gguf-py/gguf/gguf_writer.py
@@ -346,10 +346,25 @@ class GGUFWriter:
         if self.endianess == GGUFEndian.BIG:
             tensor.byteswap(inplace=True)
 
-        for fout in self.fout:
-            self.write_padding(fout, fout.tell())
-            tensor.tofile(fout)
-            self.write_padding(fout, tensor.nbytes)
+        file_id = -1
+        for i, tensors in enumerate(self.tensors):
+            if len(tensors) > 0:
+                file_id = i
+                break
+
+        fout = self.fout[file_id]
+
+        # pop the first tensor info
+        # TODO: cleaner way to get the first key
+        first_tensor_name = [name for name, _ in zip(self.tensors[file_id].keys(), range(1))][0]
+        ti = self.tensors[file_id].pop(first_tensor_name)
+        assert len(ti.shape) == len(tensor.shape)
+        assert all(dim1 == dim2 for dim1, dim2 in zip(ti.shape, tensor.shape))
+        assert ti.nbytes == tensor.nbytes
+
+        self.write_padding(fout, fout.tell())
+        tensor.tofile(fout)
+        self.write_padding(fout, tensor.nbytes)
 
         self.state = WriterState.WEIGHTS

```

---

</details>

christianazinn replied:
Rightfully so. I've forgotten how those scripts work at this point so I'll use this.

compilade reviewed (approved):
I'm satisfied with how this turned out. I did not test this extensively, but from the conversions I tried (with `--split-max-size` and with no split, both with `q8_0` and `f16`), this worked well.

A future PR to add split model support to `GGUFReader` would be nice.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -293,43 +358,57 @@ def write_tensors_to_file(self, *, progress: bool = False) -> None:
+            self.write_padding(fout, fout.tell())
 
         if self.temp_file is None:
             bar = None
+            shard_bar = None
```
Unfamiliar with how that works but yeah, probably ought to swap it for consistency.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
compatibility fix

compilade commented on gguf-py/gguf/gguf_writer.py:
```diff
@@ -282,9 +345,26 @@ def write_tensor_data(self, tensor: np.ndarray[Any, Any]) -> None:
+        # TODO: cleaner way to get the first key
+        first_tensor_name = [name for name, _ in zip(self.tensors[file_id].keys(), range(1))][0]
+        ti = self.tensors[file_id].pop(first_tensor_name)
+        assert len(ti.shape) == len(tensor.shape)
+        assert all(dim1 == dim2 for dim1, dim2 in zip(ti.shape, tensor.shape))
```
These are too strict, unfortunately.

```suggestion
```

Otherwise `q8_0` conversion with `convert-legacy-llama.py` doesn't work, and same with modifying metadata on quantized models with `gguf-new-metadata.py`.

I think this has something to do with byte shapes vs quant shapes.

At least the `nbytes` check can be kept without problems.

christianazinn replied:
Oh, this is why we used to have a `get_tensor_size` function. Not all tensors passed from `convert-legacy-llama.py` are `np.ndarray[Any, Any]`.

christianazinn committed:
Update gguf-py/gguf/gguf_writer.py

Co-authored-by: compilade <git@compilade.net>

christianazinn commented:
Forgot to mark as ready for review. Can probably be merged.

christianazinn committed:
Merge branch 'master' into convert-split

mofosyne commented:
few days has passed with the merge ready label, ci passed and approval. 

Consensus achieved? I'll presume it will be so by the end of the week.

christianazinn commented:
It's been about a week and I see no dissent so far.

compilade commented on convert-hf-to-gguf.py:
```diff
@@ -96,7 +97,8 @@ def __init__(self, dir_model: Path, ftype: gguf.LlamaFileType, fname_out: Path,
         # allow templating the file name with the output ftype, useful with the "auto" ftype
         self.fname_out = fname_out.parent / fname_out.name.format(ftype_lw, outtype=ftype_lw, ftype=ftype_lw, OUTTYPE=ftype_up, FTYPE=ftype_up)
-        self.gguf_writer = gguf.GGUFWriter(path=None, arch=gguf.MODEL_ARCH_NAMES[self.model_arch], endianess=self.endianess, use_temp_file=self.use_temp_file)
+        self.gguf_writer = gguf.GGUFWriter(None, gguf.MODEL_ARCH_NAMES[self.model_arch],endianess=self.endianess, use_temp_file=self.use_temp_file,
```
Missing space after comma, and maybe keyword args should be used

```suggestion
        self.gguf_writer = gguf.GGUFWriter(path=None, arch=gguf.MODEL_ARCH_NAMES[self.model_arch], endianess=self.endianess, use_temp_file=self.use_temp_file,
```

mofosyne replied:
If this is fixed, you happy with the overall state of the PR?

mofosyne committed:
Update convert-hf-to-gguf.py

Co-authored-by: compilade <git@compilade.net>

christianazinn committed:
Merge branch 'master' into convert-split
